File: app/src/main/java/com/majeur/psclient/service/GlobalMessageObserver.java
package com.majeur.psclient.service;

import android.util.Log;
import com.majeur.psclient.model.AvailableBattleRoomsInfo;
import com.majeur.psclient.model.BattleFormat;
import com.majeur.psclient.model.RoomInfo;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import static android.text.TextUtils.isEmpty;
import static com.majeur.psclient.util.Utils.jsonObject;

public abstract class GlobalMessageObserver extends AbsMessageObserver {

    private static final String TAG = GlobalMessageObserver.class.getSimpleName();

    private String mUserName;
    private boolean mIsUserGuest;
    private boolean mRequestServerCountsOnly;

    private final Map<String, List<String>> mPrivateMessages = new HashMap<>();

    public GlobalMessageObserver() {
        setObserveAll(true);
        mIsUserGuest = true;
    }

    public boolean isUserGuest() {
        return mIsUserGuest;
    }

    @Override
    public boolean onMessage(ServerMessage message) {
        switch (message.command) {
            case "challstr":
                processChallengeString(message);
                return true;
            case "updateuser":
                processUpdateUser(message);
                return true;
            case "nametaken":
                message.nextArg(); // Skipping name
                onShowPopup(message.nextArg());
                return true;
            case "queryresponse":
                processQueryResponse(message);
                return true;
            case "formats":
                processAvailableFormats(message);
                return true;
            case "popup":
                handlePopup(message);
                return true;
            case "updatesearch":
                handleUpdateSearch(message);
                return true;
            case "pm":
                handlePm(message);
                return true;
            case "usercount":
                return true;
            case "updatechallenges":
                handleChallenges(message);
                return true;
            case "error":
                if (message.nextArg().equals("network"))
                    onNetworkError();
                return true;
            case "init":
                onRoomInit(message.roomId, message.nextArg());
                return false; // Must not consume init/deinit commands !
            case "deinit":
                onRoomDeinit(message.roomId);
                return false; // Must not consume init/deinit commands !
            case "noinit":
                if (message.hasNextArg() && "nonexistent".equals(message.nextArg()) && message.hasNextArg())
                    onShowPopup(message.nextArg());
                return true;
            default:
                return false;
        }
    }

    private void processChallengeString(ServerMessage msg) {
        getService().putSharedData("challenge", msg.rawArgs());
        getService().tryCookieSignIn();
    }

    private void processUpdateUser(ServerMessage msg) {
        String username = msg.nextArg();
        String userType = username.substring(0, 1);
        username = username.substring(1);
        boolean isGuest = "0".equals(msg.nextArg());
        String avatar = msg.nextArg();
        avatar = ("000" + avatar).substring(avatar.length());

        mUserName = username;
        mIsUserGuest = isGuest;
        getService().putSharedData("username", username);
        onUserChanged(username, isGuest, avatar);

        // Update server counts (active battle and active users)
        mRequestServerCountsOnly = true;
        getService().sendGlobalCommand("cmd", "rooms");

        // onSearchBattlesChanged(new String[0], new String[0], new String[0]); TODO Wtf was this call ?
    }

    private void processQueryResponse(ServerMessage msg) {
        String query = msg.nextArg();
        String queryContent = msg.nextArg();
        switch (query) {
            case "rooms":
                processRoomsQueryResponse(queryContent);
                break;
            case "roomlist":
                processRoomListQueryResponse(queryContent);
                break;
            case "savereplay":
                // Not supported yet
                // queryContent is a JSON object formatted as follows:
                // {"log":"actual battle log","id":"gen7randombattle-858101987"}
                break;
            case "userdetails":
                processUserDetailsQueryResponse(queryContent);
                break;
            default:
                Log.w(TAG, "Command queryresponse not handled, type=" + query);
                break;
        }
    }

    private void processRoomsQueryResponse(String queryContent) {
        if (queryContent.equals("null"))
            return;

        try {
            JSONObject jsonObject = new JSONObject(queryContent);
            int userCount = jsonObject.getInt("userCount");
            int battleCount = jsonObject.getInt("battleCount");
            onUpdateCounts(userCount, battleCount);
            if (mRequestServerCountsOnly) {
                mRequestServerCountsOnly = false;
                return;
            }

            JSONArray jsonArray = jsonObject.getJSONArray("official");
            int N = jsonArray.length();
            RoomInfo[] officialRooms = new RoomInfo[N];
            for (int i = 0; i < N; i++) {
                JSONObject roomJson = jsonArray.getJSONObject(i);
                officialRooms[i] = new RoomInfo(roomJson.getString("title"),
                        roomJson.getString("desc"), roomJson.getInt("userCount"));
            }

            jsonArray = jsonObject.getJSONArray("chat");
            N = jsonArray.length();
            RoomInfo[] chatRooms = new RoomInfo[N];
            for (int i = 0; i < N; i++) {
                JSONObject roomJson = jsonArray.getJSONObject(i);
                chatRooms[i] = new RoomInfo(roomJson.getString("title"),
                        roomJson.getString("desc"), roomJson.getInt("userCount"));
            }

            onAvailableRoomsChanged(officialRooms, chatRooms);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private void processRoomListQueryResponse(String queryContent) {
        if (queryContent.equals("null"))
            return;

        try {

            new JSONObject(queryContent);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private void processUserDetailsQueryResponse(String queryContent) {
        try {
            JSONObject jsonObject = new JSONObject(queryContent);
            String userId = jsonObject.optString("userid");
            if (isEmpty(userId)) return;

            String name = jsonObject.optString("name");
            boolean online = jsonObject.has("status");
            String group = jsonObject.optString("group");
            Object rooms = jsonObject.opt("rooms");
            List<String> chatRooms = new LinkedList<>();
            List<String> battles = new LinkedList<>();
            if (rooms instanceof JSONObject) {
                Iterator<String> iterator = ((JSONObject) rooms).keys();
                while (iterator.hasNext()) {
                    String roomId = iterator.next();
                    if (roomId.contains("battle-")) battles.add(roomId);
                    else chatRooms.add(roomId);
                }
            }

            onUserDetails(userId, name, online, group, chatRooms, battles);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private void processAvailableFormats(ServerMessage msg) {
        String rawText = msg.rawArgs();
        List<BattleFormat.Category> battleFormatCategories = new LinkedList<>(); // /!\ needs to impl Serializable
        BattleFormat.Category currentCategory = null;
        int separator;
        while (true) {
            separator = rawText.indexOf('|');
            if (rawText.charAt(0) == ',') {
                rawText = rawText.substring(separator + 1); // Ignoring ",1|" part
                separator = rawText.indexOf('|');
                currentCategory = new BattleFormat.Category();
                battleFormatCategories.add(currentCategory);
                String categoryLabel = rawText.substring(0, separator);
                currentCategory.setLabel(categoryLabel);

            } else {
                int innerSeparator = rawText.indexOf(',');
                String formatName = rawText.substring(0, innerSeparator);
                int innerEnd = separator == -1 ? rawText.length() : separator;
                int formatInt = Integer.valueOf(rawText.substring(innerSeparator + 1, innerEnd), 16);
                BattleFormat battleFormat = new BattleFormat(formatName, formatInt);
                currentCategory.addBattleFormat(battleFormat);

            }
            if (separator == -1)
                break;
            rawText = rawText.substring(separator + 1);
        }

        getService().putSharedData("formats", battleFormatCategories);
        onBattleFormatsChanged(battleFormatCategories);
    }

    private void handlePopup(ServerMessage msg) {
        StringBuilder text = new StringBuilder(msg.nextArg());
        while (msg.hasNextArg()) {
            String next = msg.nextArg();
            if (!isEmpty(next))
                text.append("\n").append(next);
        }
        onShowPopup(text.toString());
    }

    private void handleUpdateSearch(ServerMessage msg) {
        JSONObject jsonObject = jsonObject(msg.rawArgs());
        if (jsonObject == null) return;
        String[] searching;
        JSONArray jsonArray = jsonObject.optJSONArray("searching");
        if (jsonArray == null) {
            searching = new String[0];
        } else {
            searching = new String[jsonArray.length()];
            for (int i = 0; i < searching.length; i++)
                searching[i] = jsonArray.optString(i);
        }
        JSONObject games = jsonObject.optJSONObject("games");
        String[] battleRoomIds;
        String[] battleRoomNames;
        if (games == null) {
            battleRoomIds = new String[0];
            battleRoomNames = new String[0];
        } else {
            battleRoomIds = new String[games.length()];
            battleRoomNames = new String[games.length()];
            Iterator<String> iterator = games.keys();
            for (int i = 0; iterator.hasNext(); i++) {
                String key = iterator.next();
                battleRoomIds[i] = key;
                battleRoomNames[i] = games.optString(key);
            }
        }
        onSearchBattlesChanged(searching, battleRoomIds, battleRoomNames);
    }

    private void handlePm(ServerMessage serverMessage) {
        String from = serverMessage.nextArg().substring(1);
        String to = serverMessage.nextArg().substring(1);
        String with = mUserName.equals(from) ? to : from;
        String content = serverMessage.hasNextArg() ? serverMessage.nextArg() : null;

        if (content != null && (content.startsWith("/raw") || content.startsWith("/html") || content.startsWith("/uhtml")))
            content = "Html messages not supported in pm.";
        String message = from + ": " + content;

        List<String> messages;
        if (mPrivateMessages.containsKey(with)) {
            messages = mPrivateMessages.get(with);
        } else {
            messages = new LinkedList<>();
            mPrivateMessages.put(with, messages);
        }
        messages.add(message);
        onNewPrivateMessage(with, message);
    }

    private void handleChallenges(ServerMessage message) {
        String rawJson = message.rawArgs();
        String to = null, format = null;
        String[] usersFrom = null;
        String[] formatsFrom = null;
        try {
            JSONObject jsonObject = new JSONObject(rawJson);
            if (!jsonObject.isNull("challengeTo")) {
                JSONObject challengeTo = jsonObject.getJSONObject("challengeTo");
                to = challengeTo.getString("to");
                format = challengeTo.getString("format");
            }
            JSONObject challengesFrom = jsonObject.getJSONObject("challengesFrom");
            int len = challengesFrom.length();
            usersFrom = new String[len];
            formatsFrom = new String[len];
            Iterator<String> it = challengesFrom.keys();
            for (int i = 0; i < len; i++) {
                usersFrom[i] = it.next();
                formatsFrom[i] = challengesFrom.getString(usersFrom[i]);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        if (usersFrom == null) usersFrom = formatsFrom = new String[0];
        onChallengesChange(to, format, usersFrom, formatsFrom);
    }

    @SuppressWarnings("ConstantConditions")
    public List<String> getPrivateMessages(String with) {
        if (!mPrivateMessages.containsKey(with)) return null;
        return Collections.unmodifiableList(mPrivateMessages.get(with));
    }

    protected abstract void onUserChanged(String userName, boolean isGuest, String avatarId);

    protected abstract void onUpdateCounts(int userCount, int battleCount);

    protected abstract void onBattleFormatsChanged(List<BattleFormat.Category> battleFormats);

    protected abstract void onSearchBattlesChanged(String[] searching, String[] battleRoomIds, String[] battleRoomNames);

    protected abstract void onUserDetails(String id, String name, boolean online, String group, List<String> rooms, List<String> battles);

    protected abstract void onShowPopup(String message);

    protected abstract void onAvailableRoomsChanged(RoomInfo[] officialRooms, RoomInfo[] chatRooms);

    protected abstract void onAvailableBattleRoomsChanged(AvailableBattleRoomsInfo availableRoomsInfo);

    protected abstract void onNewPrivateMessage(String with, String message);

    protected abstract void onChallengesChange(String to, String format, String[] usersFrom, String[] formatsFrom);

    protected abstract void onRoomInit(String roomId, String type);

    protected abstract void onRoomDeinit(String roomId);

    protected abstract void onNetworkError();

    public void fakeBattle() {
        getService().fakeBattle();
    }
}

File: psclient/src/main/java/com/majeur/psclient/service/observer/GlobalMessageObserver.kt
package com.majeur.psclient.service.observer

import com.majeur.psclient.model.BattleRoomInfo
import com.majeur.psclient.model.ChatRoomInfo
import com.majeur.psclient.model.common.BattleFormat
import com.majeur.psclient.service.ServerMessage
import com.majeur.psclient.service.ShowdownService
import com.majeur.psclient.util.Utils
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.util.*

class GlobalMessageObserver(service: ShowdownService)
    : AbsMessageObserver<GlobalMessageObserver.UiCallbacks>(service) {

    override var observedRoomId: String? = "lobby"
    override val interceptCommandBefore = setOf("init", "noinit")
    override val interceptCommandAfter = setOf("deinit")

    val myUsername get() = service.getSharedData<String>("myusername")?.drop(1)
    var isUserGuest: Boolean = true
        private set

    private var requestServerCountsOnly = false
    private val privateMessages = mutableMapOf<String, MutableList<String>>()

    override fun onUiCallbacksAttached() {
        // If we did not stored at least username, we will not have anything else
        val username = service.getSharedData<String>("myusername") ?: return

        onUserChanged(username, isUserGuest, /* TODO */ "000")
        onUpdateCounts(service.getSharedData("users") ?: -1,
                service.getSharedData("battles") ?: -1)

        onBattleFormatsChanged(service.getSharedData("formats") ?: emptyList())

        onSearchBattlesChanged(service.getSharedData("searching") ?: emptyList(),
                service.getSharedData("games") ?: emptyMap())

        onChallengesChange(service.getSharedData("challenge_to"),
                service.getSharedData("challenge_to_format"),
                service.getSharedData("challenge_from") ?: emptyMap())
    }

    public override fun onMessage(message: ServerMessage) {
        message.newArgsIteration()
        when (message.command) {
            "connected" -> onConnectedToServer()
            "challstr" -> processChallengeString(message)
            "updateuser" -> processUpdateUser(message)
            "queryresponse" -> processQueryResponse(message)
            "formats" -> processAvailableFormats(message)
            "popup" -> handlePopup(message)
            "updatesearch" -> handleUpdateSearch(message)
            "pm" -> handlePm(message)
            "updatechallenges" -> handleChallenges(message)
            "networkerror" -> onNetworkError()
            "init" -> onRoomInit(message.roomId, message.nextArg)
            "deinit" -> onRoomDeinit(message.roomId)
            "noinit" -> {
                if (message.hasNextArg && "nonexistent" == message.nextArg && message.hasNextArg)
                    onShowPopup(message.nextArg)
            }
            "nametaken" -> {
                message.nextArg // Skipping name
                onShowPopup(message.nextArg)
            }
            "usercount" -> {
            }
        }
    }

    private fun processChallengeString(msg: ServerMessage) {
        service.putSharedData("challenge", msg.remainingArgsRaw)
        service.tryCookieSignIn()
    }

    private fun processUpdateUser(msg: ServerMessage) {
        var username = msg.nextArg
        service.putSharedData("myusername", username)
        val userType = username.substring(0, 1)
        username = username.substring(1)
        val isGuest = "0" == msg.nextArg
        var avatar = msg.nextArg
        avatar = "000$avatar".substring(avatar.length)
        isUserGuest = isGuest
        onUserChanged(username, isGuest, avatar)

        // Update server counts (active battle and active users)
        requestServerCountsOnly = true
        service.sendGlobalCommand("cmd", "rooms")

        // onSearchBattlesChanged(new String[0], new String[0], new String[0]); TODO Wtf was this call ?
    }

    private fun processQueryResponse(msg: ServerMessage) {
        val query = msg.nextArg
        val queryResponse = msg.remainingArgsRaw
        when (query) {
            "rooms" -> processRoomsQueryResponse(queryResponse)
            "roomlist" -> processRoomListQueryResponse(queryResponse)
            "savereplay" -> processSaveReplayQueryResponse(queryResponse)
            "userdetails" -> processUserDetailsQueryResponse(queryResponse)
            else -> Timber.w("Command |queryresponse| not handled, type=$query")
        }
    }

    private fun processRoomsQueryResponse(response: String) {
        if (response == "null") return
        try {
            val jsonObject = JSONObject(response)
            val userCount = jsonObject.getInt("userCount")
            service.putSharedData("users", userCount)
            val battleCount = jsonObject.getInt("battleCount")
            service.putSharedData("battles", battleCount)
            onUpdateCounts(userCount, battleCount)
            if (requestServerCountsOnly) {
                requestServerCountsOnly = false
                return
            }
            var jsonArray = jsonObject.getJSONArray("official")
            val officialRooms = mutableListOf<ChatRoomInfo>()
            for (i in 0 until jsonArray.length()) {
                val roomJson = jsonArray.getJSONObject(i)
                officialRooms.add(
                        ChatRoomInfo(roomJson.getString("title"),
                                roomJson.getString("desc"),
                                roomJson.getInt("userCount")))
            }
            jsonArray = jsonObject.getJSONArray("chat")
            val chatRooms = mutableListOf<ChatRoomInfo>()
            for (i in 0 until jsonArray.length()) {
                val roomJson = jsonArray.getJSONObject(i)
                chatRooms.add(
                        ChatRoomInfo(roomJson.getString("title"),
                                roomJson.getString("desc"),
                                roomJson.getInt("userCount")))
            }
            onAvailableRoomsChanged(officialRooms, chatRooms)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    private fun processRoomListQueryResponse(response: String) {
        try {
            val battleRooms = mutableListOf<BattleRoomInfo>()
            val jsonObject = JSONObject(response).getJSONObject("rooms")
            val iterator = jsonObject.keys()
            while (iterator.hasNext()) {
                val roomId = iterator.next()
                val roomJson = jsonObject.getJSONObject(roomId)
                val roomInfo = BattleRoomInfo(roomId, roomJson.getString("p1"),
                        roomJson.getString("p2"), roomJson.optInt("minElo", 0))
                battleRooms.add(roomInfo)
            }
            onAvailableBattleRoomsChanged(battleRooms)
        } catch (e: JSONException) {
            e.printStackTrace()
            onAvailableBattleRoomsChanged(emptyList())
        }
    }

    private fun processSaveReplayQueryResponse(response: String) {
        try {
            val jsonObject = JSONObject(response)
            val replayId = jsonObject.optString("id")
            // val battleLog = jsonObject.optString("log")
            onReplaySaved(replayId, "https://replay.pokemonshowdown.com/$replayId")
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    private fun processUserDetailsQueryResponse(response: String) {
        try {
            val jsonObject = JSONObject(response)
            val userId = jsonObject.optString("userid")
            if (userId.isBlank()) return
            val name = jsonObject.optString("name")
            val online = jsonObject.has("status")
            val group = jsonObject.optString("group")
            val chatRooms = mutableListOf<String>()
            val battles = mutableListOf<String>()
            (jsonObject.opt("rooms") as? JSONObject)?.keys()?.forEach {
                if (it.startsWith("battle-") || it.drop(1).startsWith("battle-"))
                    battles.add(it)
                else
                    chatRooms.add(it)
            }
            onUserDetails(userId, name, online, group, chatRooms, battles)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    private fun processAvailableFormats(msg: ServerMessage) {
        val rawText: String = msg.remainingArgsRaw
        val categories: MutableList<BattleFormat.Category> = LinkedList() // /!\ needs to impl Serializable

        rawText.split("|,").forEach { rawCategory ->
            val catName = rawCategory.substringAfter("|").substringBefore("|")
            val formats = rawCategory.substringAfter(catName).split("|")
                    .filter { s -> s.isNotBlank() }
                    .map { s ->
                        BattleFormat(s.substringBefore(","), s.substringAfter(",").toInt(16))
                    }
            BattleFormat.Category().apply {
                this.formats.addAll(formats)
                label = catName
            }.also {
                categories.add(it)
            }
        }
        service.putSharedData("formats", categories)
        onBattleFormatsChanged(categories)
    }

    private fun handlePopup(msg: ServerMessage) = onShowPopup(msg.args.joinToString("\n"))

    private fun handleUpdateSearch(msg: ServerMessage) {
        val jsonObject = Utils.jsonObject(msg.remainingArgsRaw) ?: return

        val searching = mutableListOf<String>()
        val searchingJson = jsonObject.optJSONArray("searching")
        searchingJson?.let {
            for (i in 0 until searchingJson.length())
                searching.add(searchingJson.getString(i))
        }

        val games = mutableMapOf<String, String>()
        val gamesJson = jsonObject.optJSONObject("games")
        gamesJson?.let {
            gamesJson.keys().forEach { key -> games[key] = gamesJson.getString(key) }
        }
        service.putSharedData("searching", searching)
        service.putSharedData("games", games)
        onSearchBattlesChanged(searching, games)
    }

    private fun handlePm(msg: ServerMessage) {
        val from = msg.nextArg.substring(1)
        val to = msg.nextArg.substring(1)
        val myUsername = service.getSharedData<String>("myusername")?.drop(1)
        val with = if (myUsername == from) to else from
        var content = msg.nextArgSafe
        if (content != null && (content.startsWith("/raw") || content.startsWith("/html") || content.startsWith("/uhtml")))
            content = "Html messages not supported in pm."
        val message = "$from: $content"
        val messages = privateMessages.getOrPut(with, { mutableListOf<String>() })
        messages.add(message)
        onNewPrivateMessage(with, message)
    }

    private fun handleChallenges(message: ServerMessage) {
        val rawJson: String = message.remainingArgsRaw
        var to: String? = null
        var format: String? = null
        val from = mutableMapOf<String, String>()

        val jsonObject = Utils.jsonObject(rawJson)
        val challengeTo = jsonObject.optJSONObject("challengeTo")
        challengeTo?.let {
            to = it.getString("to")
            format = it.getString("format")
        }
        val challengesFrom = jsonObject.optJSONObject("challengesFrom")
        challengesFrom?.let {
            it.keys().forEach { key ->
                from[key] = challengesFrom.getString(key)
            }
        }
        service.putSharedData("challenge_to", to)
        service.putSharedData("challenge_to_format", format)
        service.putSharedData("challenge_from", from)
        onChallengesChange(to, format, from)
    }

    fun getPrivateMessages(with: String): List<String>? {
        return privateMessages[with]
    }

    fun onConnectedToServer() = uiCallbacks?.onConnectedToServer()
    fun onUserChanged(userName: String, isGuest: Boolean, avatarId: String) = uiCallbacks?.onUserChanged(userName, isGuest, avatarId)
    fun onUpdateCounts(userCount: Int, battleCount: Int) = uiCallbacks?.onUpdateCounts(userCount, battleCount)
    fun onBattleFormatsChanged(battleFormats: List<BattleFormat.Category>) = uiCallbacks?.onBattleFormatsChanged(battleFormats)
    fun onSearchBattlesChanged(searching: List<String>, games: Map<String, String>) = uiCallbacks?.onSearchBattlesChanged(searching, games)
    fun onReplaySaved(replayId: String, url: String) = uiCallbacks?.onReplaySaved(replayId, url)
    fun onUserDetails(id: String, name: String, online: Boolean, group: String, rooms: List<String>, battles: List<String>) = uiCallbacks?.onUserDetails(id, name, online, group, rooms, battles)
    fun onShowPopup(message: String) = uiCallbacks?.onShowPopup(message)
    fun onAvailableRoomsChanged(officialRooms: List<ChatRoomInfo>, chatRooms: List<ChatRoomInfo>) = uiCallbacks?.onAvailableRoomsChanged(officialRooms, chatRooms)
    fun onAvailableBattleRoomsChanged(battleRooms: List<BattleRoomInfo>) = uiCallbacks?.onAvailableBattleRoomsChanged(battleRooms)
    fun onNewPrivateMessage(with: String, message: String) = uiCallbacks?.onNewPrivateMessage(with, message)
    fun onChallengesChange(to: String?, format: String?, from: Map<String, String>) = uiCallbacks?.onChallengesChange(to, format, from)
    fun onRoomInit(roomId: String, type: String) = uiCallbacks?.onRoomInit(roomId, type)
    fun onRoomDeinit(roomId: String) = uiCallbacks?.onRoomDeinit(roomId)
    fun onNetworkError() = uiCallbacks?.onNetworkError()

    interface UiCallbacks : AbsMessageObserver.UiCallbacks {
        fun onConnectedToServer()
        fun onUserChanged(userName: String, isGuest: Boolean, avatarId: String)
        fun onUpdateCounts(userCount: Int, battleCount: Int)
        fun onBattleFormatsChanged(battleFormats: List<@JvmSuppressWildcards BattleFormat.Category>)
        fun onSearchBattlesChanged(searching: List<String>, games: Map<String, String>)
        fun onReplaySaved(replayId: String, url: String)
        fun onUserDetails(id: String, name: String, online: Boolean, group: String, rooms: List<String>, battles: List<String>)
        fun onShowPopup(message: String)
        fun onAvailableRoomsChanged(officialRooms: List<ChatRoomInfo>, chatRooms: List<ChatRoomInfo>)
        fun onAvailableBattleRoomsChanged(battleRooms: List<BattleRoomInfo>)
        fun onNewPrivateMessage(with: String, message: String)
        fun onChallengesChange(to: String?, format: String?, from: Map<String, String>)
        fun onRoomInit(roomId: String, type: String)
        fun onRoomDeinit(roomId: String)
        fun onNetworkError()
    }
}

